python/lsst/pipe/base/_instrument.py

```diff
@@ -476,3 +538,150 @@ def makeCollectionName(self, *labels: str) -> str:
+
+        This system is designed to work best when the configuration for the ID
+        packer is not overridden at all, allowing the appropriate instrument
+        class to determine the behavior for each data ID encountered.  When the
+        configuration does need to be modified (most often when the scheme for
+        packing an instrument's data IDs is undergoing an upgrade), it is
+        important to ensure the overrides are only applied to data IDs with the
+        desired instrument value.
+
+        Unit tests of code that use a field produced by this method will often
+        want to explicitly set the packer to "observation" and manually set
+        its ``n_detectors`` and ``n_observations`` fields; this will make it
+        unnecessary for tests to provide expanded data IDs.
+        """
+        # The control flow here bounces around a bit when this RegistryField's
+        # apply() method is called, so it merits a thorough walkthrough
+        # somewhere, and that might as well be here:
+        #
+        # - If the config field's name is not `None`, that kind of packer is
+        #   constructed and returned with the arguments to `apply`, in just
+        #   the way it works with most RegistryFields or ConfigurableFields.
+        #   But this is expected to be rare.
+        #
+        # - If the config fields' name is `None`, the `apply` method (which
+        #   actually lives on the `pex.config.RegistryInstanceDict` class,
+        #   since `RegistryField` is a descriptor), calls
+        #   `_make_default_dimension_packer_dispatch` (which is final, and
+        #   hence the base class implementation just below is the only one).
+        #
+        # - `_make_default_dimension_packer_dispatch` instantiates an
+        #   `Instrument` instance of the type pointed at by the data ID
+        #   (i.e. calling `Instrument.from_data_id`), then calls
+        #   `_make_default_dimension_packer` on that.
+        #
+        # - The default implementation of `_make_default_dimension_packer`
+        #    here in the base class picks the "observation" dimension packer,
+        #   so if it's not overridden by a derived class everything proceeds
+        #   as if the config field's name was set to that.  Note that this
+        #   sets which item in the registry is used, but it still pays
+        #   attention to the configuration for that entry in the registry
+        #   field.
+        #
+        # - A subclass implementation of `_make_default_dimension_packer` will
+        #   take precedence over the base class, but it's expected that these
+        #   will usually just delegate back to ``super()`` while changing
+        #   the ``default`` argument to something other than "observation".
+        #   Once again, this will control which packer entry in the registry is
+        #   but the result will still reflect the configuration for that packer
+        #   in the registry field.
```

Comment:
I'm torn on whether this long comment deserves to be in the Notes section above, or not. Comments that are this long often suggest to me that they should be proper docs, but this really is an implementation detail. It also took me a few readings to catch all the nuances, but there's probably nothing to be done about that.